### PR TITLE
Adds size-specific mAP metrics (`map_small`, `map_medium`, `map_large`) to the `Metric` class

### PR DESCRIPTION
**Pull request by ashish-reddy26**
I have read the CLA Document and I sign the CLA

This PR addressess the issue #20974 (references #6043)

## Description 

Adds size-specific mAP metrics (`map_small`, `map_medium`, `map_large`) to the `Metric` class, making them programmatically accessible through `Metric.box.map_small`, `Metric.box.map_medium`, and `Metric.box.map_large`.

## Changes Made

- **Added `size_specific_metrics` list** to `Metric` class initialization with default values `[0.0, 0.0, 0.0]`
- **Added three new properties** to `Metric` class:
  - `map_small`: Returns mAP for small objects (IoU 0.5-0.95)
  - `map_medium`: Returns mAP for medium objects (IoU 0.5-0.95)  
  - `map_large`: Returns mAP for large objects (IoU 0.5-0.95)
- **Added `update_size_metrics()` method** for updating values from COCO evaluation
- **Modified `eval_json()` in `DetectionValidator`** to capture COCO evaluator stats[3:6] and populate size-specific metrics
- **Added comprehensive tests** in `tests/test_size_metrics.py` covering:
  - Unit tests for `Metric` class functionality
  - Integration tests for COCO and non-COCO scenarios
  - Property validation and update mechanisms

## Testing

- [x] **COCO datasets**: Size-specific metrics populate with real values from COCO evaluator
- [x] **Non-COCO datasets**: Metrics default to 0.0 as expected
- [x] **Unit tests**: All 5 tests pass, covering initialization, updates, and properties
- [x] **Integration tests**: COCO validation pipeline works correctly  
- [x] **Backward compatibility**: Existing metrics unchanged and working
- [x] **API verification**: Confirmed access via `results.box.map_small`, `results.box.map_medium`, `results.box.map_large`

## Usage Example

```python
from ultralytics import YOLO

# Load model and train 
model = YOLO("yolo11n.pt")
results = model.train(data="coco.yaml")

# Access size-specific metrics (COCO datasets)
print(f"Small objects mAP: {results.box.map_small:.3f}")
print(f"Medium objects mAP: {results.box.map_medium:.3f}")  
print(f"Large objects mAP: {results.box.map_large:.3f}")

# For non-COCO datasets, values default to 0.0
results_yolo = model.val(data="custom_yolo_dataset.yaml")
print(f"Small objects mAP: {results_yolo.box.map_small}")  # 0.0
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds size-specific mAP metrics (small/medium/large) to YOLO detection validation, exposing `map_small`, `map_medium`, and `map_large` when evaluating on COCO datasets. 📏📦

### 📊 Key Changes
- New size-specific metrics in `Metric`:
  - Properties: `map_small`, `map_medium`, `map_large`
  - Storage: `size_specific_metrics = [0.0, 0.0, 0.0]`
  - Method: `update_size_metrics(map_small, map_medium, map_large)`
- COCO integration in `DetectionValidator.eval_json`:
  - Automatically extracts and updates size-specific AP values from COCO evaluation when available.
- Tests added:
  - Unit tests for initialization, updates, and property correctness.
  - Integration test (COCO-only, requires `pycocotools`) validating metrics are populated.
  - Default behavior test ensuring non-COCO datasets return 0.0 for size metrics.

### 🎯 Purpose & Impact
- Better insights: Users can now analyze model performance by object size (small/medium/large) on COCO. 🔍
- Backward compatible: No breaking API changes; defaults remain 0.0 for non-COCO datasets.
- Easier benchmarking: Facilitates fine-grained evaluation and model comparison, especially for datasets with varied object scales.
- Quick usage example:
  ```python
  from ultralytics import YOLO

  model = YOLO("yolo11n.pt")
  results = model.val(data="coco8.yaml", imgsz=640)
  print(results.metrics.box.map_small, results.metrics.box.map_medium, results.metrics.box.map_large)
  ```
  Note: Size metrics are populated for COCO; non-COCO datasets will return 0.0.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @AshishreddyM26, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process. An Ultralytics engineer will assist you soon.

Please review the following checklist to ensure your contribution is ready for integration:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀🦾

**Comment by ashish-reddy26** [comment]
I have read the CLA Document and I sign the CLA

**Comment by RizwanMunawar** [comment]
@AshishreddyM26 Hi, thanks for the PR. We will review it soon, Please note that it may take some time. @Laughing-q, maybe we can review this. Thanks

**Comment by Laughing-q** [comment]
@AshishreddyM26 hey thanks for the PR! some review comments I dropped here:
- let's not have `size_specific_metrics` as an intermediate attribute and directly add `map_small/map_medium/map_large` attributes in Metrics
- also for this probably we don't want a specific test for this, can you eliminate that part?
- if possible can you also add small/medium/large mAPs for masks/pose tasks?

Thank you!
